### PR TITLE
First proposal to add room capture

### PR DESCRIPTION
**Pull request by cabanier**
/agenda give browser the ability to start a room capture

***
<a href="https://pr-preview.s3.amazonaws.com/cabanier/real-world-geometry/pull/35.html" title="Last updated on Oct 27, 2022, 11:14 PM UTC (8fb22dd)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/immersive-web/real-world-geometry/35/3776bf2...cabanier:8fb22dd.html" title="Last updated on Oct 27, 2022, 11:14 PM UTC (8fb22dd)">Diff</a>

### CODE REVIEW

**Comment by cabanier** [comment]
> Have a conflict, so I can't join for today's discussion!
> 
> I am not sure what it means to "start a room capture" from this new API as compared to the normal operation of the plane-finding API. Can you link to any docs describing how this feature works today in OpenXR or as a Quest user experience?

Quest does not have the ability to find planes. Instead it relies on the user to set up the room.
Here's a video on what that looks like: https://www.youtube.com/watch?v=4t1CdmDeBhA
Basically, if no planes are found, an author can prompt the user if they want to run this procedure.

We discussed this API a bit during today's meeting. @bialpio thought that this could be use on Android to initiate the plane finding (after which the promise returned). Alternatively, we will have to differentiate between systems where the planes are always known (Quest) vs where they have to be discovered each time (Android).
I'm unsure what Hololens' behavior is here.

**Comment by cabanier** [comment]
I would expect that this API returns immediately if your device doesn't support an interactive room capture.

**Comment by cabanier** [comment]
/tpac discuss method to allow room capture

**Comment by cabanier** [comment]
I hooked up this API as an experiment and the experience is much nicer than what we have today: https://www.youtube.com/shorts/hNAj5gL7m60
We recently shipped planes and I've had to tell many people that they need to go to settings to get the planes to work. It will be much better if the experience itself can trigger this
/agenda add API to trigger room setup

**Comment by cabanier** [comment]
@thetuvix the YouTube link shows how the feature works.
@bialpio do you think we can pull this in?

**Comment by bialpio** [comment]
I think the biggest question I have here is, what is the intended flow for this API? The use case to me seems to be that if a user enters XRSession w/o having configured a room / playable area prior to using WebXR, this gives the site a chance to trigger the system UI to do so, correct? Is there ever a case where a site would want to trigger it in the middle of an experience? Is this something that the UA could automatically trigger prior to entering the session, in a manner that is opaque to the site?

**Comment by cabanier** [comment]
> I think the biggest question I have here is, what is the intended flow for this API? 

Take a look at the youtube video I posted.
> The use case to me seems to be that if a user enters XRSession w/o having configured a room / playable area prior to using WebXR, this gives the site a chance to trigger the system UI to do so, correct?

Yes. I imagine that is the case *most* of the time

> Is there ever a case where a site would want to trigger it in the middle of an experience? Is this something that the UA could automatically trigger prior to entering the session, in a manner that is opaque to the site?

I can imagine an experience that starts off in a more general matter and then wants to transition to a scene that requires world planes.
Also, some experience might want to ask the user if they want planes or just want to jump into a more generic experience.

**Comment by cabanier** [comment]
@bialpio I marked it as unstable. Can we pull it in so we can iterate on it?

**Comment by bialpio** [comment]
> > I think the biggest question I have here is, what is the intended flow for this API?
> 
> Take a look at the youtube video I posted.
> 
> > The use case to me seems to be that if a user enters XRSession w/o having configured a room / playable area prior to using WebXR, this gives the site a chance to trigger the system UI to do so, correct?
> 
> Yes. I imagine that is the case _most_ of the time
> 

Gotcha, I think that is mostly why I'm confused - why can't the XR system ensure that planes are set up when a site requests a session w/ "plane-detection" feature listed as required / optional? I.e. why is the site even involved in making this call?

> > Is there ever a case where a site would want to trigger it in the middle of an experience? Is this something that the UA could automatically trigger prior to entering the session, in a manner that is opaque to the site?
> 
> I can imagine an experience that starts off in a more general matter and then wants to transition to a scene that requires world planes. Also, some experience might want to ask the user if they want planes or just want to jump into a more generic experience.

I feel like we may be able to achieve the transition from more general to specific experience by ensuring that the site can ask for additional features at runtime (immersive-web/webxr#1190) & letting the XR system ensure that it has gone through an appropriate setup to make the requested features work. IIUC, w/o the extra call, on some systems we'll have a situation where a feature is enabled, but nothing gets returned because the system wasn't configured?

In any case, I don't feel strongly enough to block this PR from landing (LMK if you need me to press some button somewhere), but please make sure to bring it up to the rest of the group as I think it'd be the first time we're introducing this kind of pattern - my understanding is that we generally tried not to introduce a "configure this API" methods post-session-creation, but may be misremembering.

**Review by bialpio — APPROVED** [review]
Approving so that we can iterate.

**Comment by cabanier** [comment]
> > Yes. I imagine that is the case _most_ of the time
> 
> Gotcha, I think that is mostly why I'm confused - why can't the XR system ensure that planes are set up when a site requests a session w/ "plane-detection" feature listed as required / optional? I.e. why is the site even involved in making this call?

The manual room setup will always result in the highest quality plane information.
Even if a future OS update starts automatically finding planes, sites might still prefer if users mark up their room manually.
Worst case, the user immediately cancel room setup and the site has to deal with it.

> > I can imagine an experience that starts off in a more general matter and then wants to transition to a scene that requires world planes. Also, some experience might want to ask the user if they want planes or just want to jump into a more generic experience.
> 
> I feel like we may be able to achieve the transition from more general to specific experience by ensuring that the site can ask for additional features at runtime ([immersive-web/webxr#1190](https://github.com/immersive-web/webxr/issues/1190)) & letting the XR system ensure that it has gone through an appropriate setup to make the requested features work. IIUC, w/o the extra call, on some systems we'll have a situation where a feature is enabled, but nothing gets returned because the system wasn't configured?

I agree.
Sites are going to have to present user with a choice:
- mark up your room for best experience
- rely on auto-detection of planes (like on Android) to get an OK-ish experience. (Tell the user to look around first?)
- not rely on planes at all

**Pull request merged**
